### PR TITLE
Avoid null values when resolving alternative priority

### LINKED ISSUE #8895: NPE Comparing alternative priorities
**Describe the bug**
`Beans.compareAlternativeBeans` always prefers the value of the declaring bean, even if the declaring bean doesn't have an alternative priority and the bean itself does. This leads to an NPE when comparing alternate priorities.

**Expected behavior**
I expect something like this to work be resolvable (not sure this exactly reproduces.. I'm working on it)

```
    @Dependent
    static class NoParentAlternativePriority {

        @Produces
        @ApplicationScoped
        @AlternativePriority(3)
        PrioritizedProducerMethodBean bar = new PrioritizedProducerMethodBean();

        @Produces
        @ApplicationScoped
        @AlternativePriority(4)
        public PrioritizedProducerMethodBean createBar() {
            return new PrioritizedProducerMethodBean();
        }
    }
```

**Actual behavior**
As it happened in my situation, the comparison resulted in an NPE: 

```[ERROR] dev.ebullient.micrometer.deployment.export.SecondPrometheusTest  Time elapsed: 0.354 s  <<< ERROR!
java.lang.RuntimeException: 
java.lang.RuntimeException: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
        [error]: Build step io.quarkus.arc.deployment.ArcProcessor#validate threw an exception: java.lang.NullPointerException
        at io.quarkus.arc.processor.Beans.compareAlternativeBeans(Beans.java:551)
        at java.base/java.util.TimSort.countRunAndMakeAscending(TimSort.java:355)
        at java.base/java.util.TimSort.sort(TimSort.java:220)
        at java.base/java.util.Arrays.sort(Arrays.java:1515)
        at java.base/java.util.ArrayList.sort(ArrayList.java:1749)
        at io.quarkus.arc.processor.Beans.resolveAmbiguity(Beans.java:523)
        at io.quarkus.arc.processor.Beans.resolveInjectionPoint(Beans.java:475)
        at io.quarkus.arc.processor.BeanInfo.init(BeanInfo.java:420)
        at io.quarkus.arc.processor.BeanDeployment.init(BeanDeployment.java:224)
        at io.quarkus.arc.processor.BeanProcessor.initialize(BeanProcessor.java:130)
        at io.quarkus.arc.deployment.ArcProcessor.validate(ArcProcessor.java:291)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at io.quarkus.deployment.ExtensionLoader$2.execute(ExtensionLoader.java:938)
        at io.quarkus.builder.BuildContext.run(BuildContext.java:273)
        at org.jboss.threads.ContextClassLoaderSavingRunnable.run(ContextClassLoaderSavingRunnable.java:35)
        at org.jboss.threads.EnhancedQueueExecutor.safeRun(EnhancedQueueExecutor.java:2027)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.doRunTask(EnhancedQueueExecutor.java:1551)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1442)
        at java.base/java.lang.Thread.run(Thread.java:834)
        at org.jboss.threads.JBossThread.run(JBossThread.java:479)
```

Will create a PR to fix the problem that includes a reproducing test.

### PR DESCRIPTION
**Pull request by ebullient**
Avoid comparing null values when comparing alternative priorities
Resolves #8895

### CODE REVIEW

**Comment by geoand** [comment]
Interesting!

**Comment by manovotn** [comment]
@ebullient if I take just the test and run that on current master (without the changes done to `Beans` class in this PR), I still get all tests to pass. So the test isn't manifesting the problem at all.

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/Beans.java [hunk 1]
```diff
@@ -558,10 +558,24 @@ private static Integer getAlternativePriority(BeanInfo bean) {
...
+        if (priority2 == null) {
+            priority2 = bean2.getDeclaringBean().getAlternativePriority();
+        }
+
+        priority1 = bean1.getAlternativePriority();
+        if (priority1 == null) {
+            priority1 = bean1.getDeclaringBean().getAlternativePriority();
+        }
+
+        if (priority2 == null) {
```
Comment (mkouba) [inline independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/Beans.java:573]: At this point `priority1` and `priority2` may not be null. So I'd either remove the checks or throw an ISE or something like that.

**Comment by mkouba** [comment]
@ebullient @manovotn The test does not demonstrate the issue because it's using programmatic lookup (`Instance<PrioritizedProducerMethodBean>`) and so the resolution is performed at runtime - `io.quarkus.arc.impl.ArcContainerImpl.compareAlternativeBeans(InjectableBean<?>, InjectableBean<?>)` works correctly.

In order to test `Beans.compareAlternativeBeans(BeanInfo, BeanInfo)` you would have to inject the `PrioritizedProducerMethodBean` in a bean. I think we should extract the test into a separate test class to make it clear.

**Comment by manovotn** [comment]
Thanks @mkouba, that explains it :)

I think it would be useful to re-use all these tests in both cases, runtime and build time (the test now only uses `Instance` all the way). That could be easily done by adding one more bean into the mix that simply injects all the beans and doing some more assertions on that.

**Comment by ebullient** [comment]
> Thanks @mkouba, that explains it :)
> 
> I think it would be useful to re-use all these tests in both cases, runtime and build time (the test now only uses `Instance` all the way). That could be easily done by adding one more bean into the mix that simply injects all the beans and doing some more assertions on that.

I can move to a different class and fix the test. I know it works because it fixes my code ... but getting it to fail in this environment was a bit magical for me. Will try again. Thanks. ;)

**Comment by ebullient** [comment]
> I can move to a different class and fix the test. I know it works because it fixes my code ... but getting it to fail in this environment was a bit magical for me. Will try again. Thanks. ;)

Moved the test to a different class. Undid the patch, and confirmed exactly the same exception that I observed that lead me down this road to begin with. =)

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/Beans.java (hunk 1 above)
Comment (ebullient) [reply independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/Beans.java:573]: I can change to ISE. Paranoid as getting here in the first place was unexpected. ;)

Comment (manovotn) [reply independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/Beans.java:573]: ISE sounds good, +1

**Comment by manovotn** [comment]
@ebullient test looks good now, thanks for spotting && fixing this bug! :)

**Review by mkouba — APPROVED** [review]

**Pull request merged**